pkg/vmcp/server/adapter/capability_adapter.go

```diff
@@ -149,3 +149,57 @@ func (a *CapabilityAdapter) ToSDKPrompts(prompts []vmcp.Prompt) []server.ServerP
 
 	return sdkPrompts
 }
+
+// ToCompositeToolSDKTools converts composite tools to SDK ServerTool format with workflow handlers.
+//
+// This method is similar to ToSDKTools but uses composite tool workflow handlers instead of
+// backend routing handlers. For each composite tool:
+//   - Marshals InputSchema to JSON (SDK expects RawInputSchema as []byte)
+//   - Creates workflow handler via HandlerFactory.CreateCompositeToolHandler
+//   - Wraps in server.ServerTool struct
+//
+// The workflowExecutors map provides the workflow executor for each tool name.
+// Returns error if schema marshaling fails or workflow executor is missing for any tool.
+func (a *CapabilityAdapter) ToCompositeToolSDKTools(
+	tools []vmcp.Tool,
+	workflowExecutors map[string]WorkflowExecutor,
+) ([]server.ServerTool, error) {
+	if len(tools) == 0 {
+		return nil, nil
+	}
+
+	sdkTools := make([]server.ServerTool, 0, len(tools))
+	for _, tool := range tools {
+		// Get workflow executor for this tool
+		executor, exists := workflowExecutors[tool.Name]
```

Comment:
I think this is interesting because we are creating the composite tool in the per-session scope (which I think is correct) based on session-discovered tools. I wonder if we can ever run into the situation where the workflow will include tools that not all users have access to..in which case failing hard might not be the best option, we might rather "disable" the workflow.

Not for this PR I guess but we should keep it in mind..maybe add a comment or proactively return a special error and handle it at the upper layer in a very simple way (just popping up and log).